### PR TITLE
[CIT-249] Refactor command trait for mqtt

### PR DESCRIPTION
**Pull request by rina23q**
## Proposed changes
This PR a part of the series command trait refactoring.
- Make clear separation between command, cli option, and error.
- Rename the former structs.
- Introduce a struct specific to each sub command to fit the build_command.
- Organise the code with one file per command.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: tedge/src/cli/mqtt/publish.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+#[tokio::main]
+async fn publish(cmd: &MqttPublishCommand) -> Result<(), MqttError> {
+    let mut mqtt = cmd
+        .mqtt_config
+        .clone()
+        .connect(cmd.client_id.as_str())
+        .await?;
+
+    let tpc = Topic::new(cmd.topic.as_str())?;
```
Comment (mneumann) [inline tedge/src/cli/mqtt/publish.rs:38]: change varible name to `topic`. Also think about changing the type of `topic` in MqttPublishCommand from `String` to `Topic`.

File: tedge/src/cli/mqtt/publish.rs [hunk 2]
```diff
@@ -0,0 +1,75 @@
+use crate::cli::mqtt::MqttError;
+use crate::command::{Command, ExecutionContext};
+use futures::future::FutureExt;
+use mqtt_client::{Client, Message, QoS, Topic};
+use std::time::Duration;
+use tokio::select;
+
+pub struct MqttPublishCommand {
+    pub topic: String,
```
Comment (mneumann) [inline tedge/src/cli/mqtt/publish.rs:9]: change type of `topic` to `mqtt_client::Topic` and do the "parsing" of the topic in the BuildCommand.

File: tedge/src/cli/mqtt/subscribe.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+use crate::command::{Command, ExecutionContext};
+use crate::utils::signals;
+use futures::future::FutureExt;
+use mqtt_client::{Client, Message, MessageStream, QoS, TopicFilter};
+use tokio::{io::AsyncWriteExt, select};
+
+pub struct MqttSubscribeCommand {
+    pub topic: String,
+    pub qos: QoS,
+    pub no_topic: bool,
```
Comment (mneumann) [inline tedge/src/cli/mqtt/subscribe.rs:11]: what is the meaning of `no_topic`? Try to find a better way. I have to look at the source code to understand the meaning of it.

Comment (mneumann) [reply tedge/src/cli/mqtt/subscribe.rs:11]: `print_topic`, `include_topic_in_output`? Try to use positive logic.

Comment (mneumann) [reply tedge/src/cli/mqtt/subscribe.rs:11]: `hide_topic` maybe. I now saw that it refers to the CLI option `--no-topic`. But in the code it's really hard to understand it's meaning

**Review by mneumann — CHANGES_REQUESTED** [review]
Looks good so far to me. Some minor changes.

File: tedge/src/cli/mqtt/subscribe.rs (hunk 3 above)
Comment (rina23q) [reply tedge/src/cli/mqtt/subscribe.rs:11]: Changed to `hide_topic`

**Review by mneumann — APPROVED** [review]
Nice

File: tedge/src/cli/mqtt/cli.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    match int_val {
+        0 => Ok(QoS::AtMostOnce),
+        1 => Ok(QoS::AtLeastOnce),
+        2 => Ok(QoS::ExactlyOnce),
+        _ => Err(MqttError::InvalidQoSError),
+    }
+}
+
+#[test]
+fn test_parse_qos() {
```
Comment (albinsuresh) [inline tedge/src/cli/mqtt/cli.rs:89]: Should't this test be in a test module?

File: tedge/src/cli/mqtt/cli.rs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    let input_qos = "0";
```
Comment (albinsuresh) [inline tedge/src/cli/mqtt/cli.rs:90]: Unrelated to this PR:The test coverage could be improved with validations for other QOS values as well.

**Review by albinsuresh — CHANGES_REQUESTED** [review]

File: tedge/src/cli/mqtt/cli.rs (hunk 4 above)
Comment (rina23q) [reply tedge/src/cli/mqtt/cli.rs]: I think yes. But it worked without in the "tests" module. Why? :(

File: tedge/src/cli/mqtt/cli.rs (hunk 5 above)
Comment (rina23q) [reply tedge/src/cli/mqtt/cli.rs]: I added more tests.

**Review by albinsuresh — APPROVED** [review]

**Pull request merged**